File: leetcode/dp/2189_Number_of_Ways_to_Build_House_of_Cards/codes/dp(ver2).py
dp = [0] * 501
dp[0] = 1
for row_size in range(2, 501, 3):
    for reachable in range(500, row_size - 1, -1):
        dp[reachable] += dp[reachable - row_size]

class Solution:
    def houseOfCards(self, n: int) -> int:
        return dp[n]

File: leetcode/dp/2189_Number_of_Ways_to_Build_House_of_Cards/codes/dp(ver2).c
#include <stdio.h>

int houseOfCards(int n) {
    static int dp[501];
    static int initialized = 0;
    if (!initialized) {
        dp[0] = 1;
        for (int row_size = 2; row_size <= 500; row_size += 3) {
            for (int reachable = 500; reachable >= row_size; reachable--) {
                dp[reachable] += dp[reachable - row_size];
            }
        }
        initialized = 1;
    }
    return dp[n];
}
